File: rs_ai/backend/services/nib_service.py
import os
import json
import requests
from datetime import datetime
from .cache_service import get_cached_data, set_cached_data

# Initialize OpenRouter client
openrouter_api_key = os.environ.get('OPENROUTER_API_KEY')

# Use the specified model
MODEL = "mistralai/mistral-small-3.1-24b-instruct:free"

def get_nib_recommendations():
    """
    Get NIB recommendations
    
    Returns:
        NIB recommendations
    """
    cache_key = "nib_recommendations"
    cached_data = get_cached_data(cache_key, 7200)  # Cache for 2 hours
    
    if cached_data:
        return cached_data
    
    try:
        # Get basic NIB info
        basic_info = get_nib_basic_info()
        
        # Get AI recommendations
        sustainable_recommendation = generate_ai_recommendation(
            "Sustainable Finance",
            """Generate an AI investment recommendation for sustainable finance projects in Nordic and Baltic regions.
            Focus on green finance, renewable energy, and sustainable infrastructure projects."""
        )
        
        infrastructure_recommendation = generate_ai_recommendation(
            "Infrastructure Development",
            """Generate an AI investment recommendation for infrastructure development projects in Nordic and Baltic regions.
            Focus on transportation, energy networks, and digital infrastructure."""
        )
        
        innovation_recommendation = generate_ai_recommendation(
            "Innovation Finance",
            """Generate an AI investment recommendation for innovation finance projects in Nordic and Baltic regions.
            Focus on technology startups, research and development, and digital transformation."""
        )
        
        # Combine into final result
        result = {
            "basic": basic_info,
            "aiRecommendations": {
                "sustainable": sustainable_recommendation,
                "infrastructure": infrastructure_recommendation,
                "innovation": innovation_recommendation
            },
            "analysisDate": datetime.now().strftime("%Y-%m-%d"),
            "modelDisclaimer": f"Recommendations generated using {MODEL}. These are AI-generated suggestions for informational purposes only and should not be considered financial advice."
        }
        
        # Cache the result
        set_cached_data(cache_key, result, 7200)  # Cache for 2 hours
        
        return result
    except Exception as e:
        print(f"Error getting NIB recommendations: {str(e)}")
        return create_fallback_nib_recommendations()

def get_nib_basic_info():
    """
    Get basic NIB information
    
    Returns:
        Basic NIB information
    """
    return {
        "yearInBrief": {
            "title": "Nordic Investment Bank - Year in Brief",
            "description": "The Nordic Investment Bank (NIB) is an international financial institution owned by eight member countries: Denmark, Estonia, Finland, Iceland, Latvia, Lithuania, Norway and Sweden. NIB finances private and public projects in and outside the member countries.",
            "keyHighlights": [
                {
                    "title": "Loans Disbursed",
                    "value": "EUR 4.2 billion",
                    "description": "Total amount of loans disbursed in the previous year",
                    "sourceUrl": "https://www.nib.int/financial-report"
                },
                {
                    "title": "New Agreements Signed",
                    "value": "39",
                    "description": "Number of new loan agreements signed",
                    "sourceUrl": "https://www.nib.int/loans/agreed-loans"
                },
                {
                    "title": "Green Bond Issuance",
                    "value": "EUR 1.5 billion",
                    "description": "Total green bonds issued",
                    "sourceUrl": "https://www.nib.int/investors/funding/funding-programmes"
                }
            ],
            "factSheets": [
                {
                    "title": "Impact Report",
                    "description": "Annual impact report focusing on climate and society impacts",
                    "link": "https://www.nib.int/releases/nib-s-annual-impact-report-2023-addressing-climate-change-and-supporting-sustainable-growth"
                },
                {
                    "title": "Financial Report",
                    "description": "Annual financial report with key performance indicators",
                    "link": "https://www.nib.int/financial-report"
                },
                {
                    "title": "Sustainability Policy",
                    "description": "NIB's approach to sustainable financing",
                    "link": "https://www.nib.int/sustainability"
                }
            ]
        },
        "quickLinks": [
            {
                "title": "Loan Products",
                "icon": "currency-dollar",
                "description": "Explore our loan products and services",
                "link": "https://www.nib.int/loans"
            },
            {
                "title": "Investment Criteria",
                "icon": "check-circle",
                "description": "Learn about our investment criteria and focus areas",
                "link": "https://www.nib.int/loans/loan-products"
            },
            {
                "title": "Application Process",
                "icon": "file-text",
                "description": "Step-by-step guide to the loan application process",
                "link": "https://www.nib.int/loans/how-to-borrow"
            },
            {
                "title": "Sustainable Investments",
                "icon": "leaf",
                "description": "Our approach to sustainable financing",
                "link": "https://www.nib.int/sustainability"
            }
        ]
    }

def generate_ai_recommendation(sector, prompt):
    """
    Generate AI recommendation for a specific sector
    
    Args:
        sector: Investment sector
        prompt: Prompt for the AI model
        
    Returns:
        AI recommendation
    """
    try:
        # Enhance the prompt
        full_prompt = f"""
You are an investment advisor for the Nordic Investment Bank. Your task is to provide a detailed
investment recommendation for {sector} in the Nordic and Baltic regions.

{prompt}

Format your response as a valid JSON object with the following fields:
1. "title": A concise, specific project title (e.g., "Green Hydrogen Production in Denmark")
2. "description": A one-sentence project description
3. "industry": The specific industry within {sector}
4. "riskLevel": Risk assessment ("Low", "Medium", or "High")
5. "opportunityLevel": Opportunity assessment ("Low", "Medium", or "High")
6. "analysis": A detailed 120-150 word analysis of market conditions, risks, and opportunities
7. "keyRecommendation": A specific, actionable investment recommendation including suggested amount

The response should be a valid JSON object containing only these fields.
"""
        
        headers = {
            "Authorization": f"Bearer {openrouter_api_key}",
            "Content-Type": "application/json"
        }
        
        data = {
            "model": MODEL,
            "messages": [
                {"role": "user", "content": full_prompt}
            ],
            "response_format": {"type": "json_object"}
        }
        
        response = requests.post(
            "https://openrouter.ai/api/v1/chat/completions",
            headers=headers,
            json=data
        )
        
        if response.status_code == 200:
            response_data = response.json()
            content = response_data.get("choices", [{}])[0].get("message", {}).get("content", "")
            
            try:
                # Parse as JSON
                recommendation = json.loads(content)
                
                # Validate required fields
                required_fields = ["title", "description", "industry", "riskLevel", "opportunityLevel", "analysis", "keyRecommendation"]
                for field in required_fields:
                    if field not in recommendation:
                        recommendation[field] = f"Missing {field}"
                
                return recommendation
            except json.JSONDecodeError:
                print(f"Invalid JSON from AI: {content}")
                return create_fallback_recommendation(sector)
        else:
            print(f"Error from OpenRouter API: {response.status_code} - {response.text}")
            return create_fallback_recommendation(sector)
    except Exception as e:
        print(f"Error generating AI recommendation: {str(e)}")
        return create_fallback_recommendation(sector)

def create_fallback_recommendation(sector):
    """
    Create fallback recommendation if AI fails
    
    Args:
        sector: Investment sector
        
    Returns:
        Fallback recommendation
    """
    sector_titles = {
        "Sustainable Finance": "Green Energy Investment Fund",
        "Infrastructure Development": "Urban Transportation Modernization",
        "Innovation Finance": "Digital Healthcare Initiative"
    }
    
    return {
        "title": sector_titles.get(sector, f"{sector} Initiative"),
        "description": f"A strategic investment in {sector} across Nordic and Baltic regions",
        "industry": sector,
        "riskLevel": "Medium",
        "opportunityLevel": "High",
        "analysis": f"The {sector} sector in Nordic and Baltic countries presents significant growth opportunities due to supportive government policies, strong institutional frameworks, and innovative business environment. While there are challenges related to market competition and regulatory complexity, the long-term outlook remains positive due to increasing demand for sustainable solutions and digital transformation.",
        "keyRecommendation": f"Invest 20-25% of portfolio in {sector} projects with focus on cross-border collaborations and public-private partnerships."
    }

def create_fallback_nib_recommendations():
    """
    Create fallback NIB recommendations if the entire process fails
    
    Returns:
        Fallback NIB recommendations
    """
    return {
        "basic": get_nib_basic_info(),
        "aiRecommendations": {
            "sustainable": create_fallback_recommendation("Sustainable Finance"),
            "infrastructure": create_fallback_recommendation("Infrastructure Development"),
            "innovation": create_fallback_recommendation("Innovation Finance")
        },
        "analysisDate": datetime.now().strftime("%Y-%m-%d"),
        "modelDisclaimer": f"Recommendations generated using {MODEL}. These are AI-generated suggestions for informational purposes only and should not be considered financial advice."
    }

File: server/services/nibService.ts
// Define types for OpenRouter API
export interface OpenRouterMessage {
  role: "system" | "user" | "assistant";
  content: string;
}
import fetch from "node-fetch";

// Dynamic import for cache service to match the server pattern
async function getCacheService() {
  return await import('./cacheService');
}

// Define interface for OpenRouter response
interface OpenRouterResponse {
  id: string;
  object: string;
  created: number;
  model: string;
  choices: {
    index: number;
    message: {
      role: string;
      content: string;
    };
    finish_reason: string;
  }[];
  usage: {
    prompt_tokens: number;
    completion_tokens: number;
    total_tokens: number;
  };
}

// Cache keys
const NIB_RECOMMENDATIONS_CACHE_KEY = "nib_recommendations_";

// NIB data structure
export interface NIBBasicInfo {
  yearInBrief: {
    title: string;
    description: string;
    keyHighlights: {
      title: string;
      value: string;
      description: string;
      sourceUrl: string;
    }[];
    factSheets: {
      title: string;
      description: string;
      link: string;
    }[];
  };
  quickLinks: {
    title: string;
    icon: string;
    description: string;
    link: string;
  }[];
}

export interface AIRecommendation {
  title: string;
  description: string;
  industry: string;
  riskLevel: string;
  opportunityLevel: string;
  analysis: string;
  keyRecommendation: string;
}

export interface NIBRecommendations {
  basic: NIBBasicInfo;
  aiRecommendations: {
    sustainable: AIRecommendation;
    infrastructure: AIRecommendation;
    innovation: AIRecommendation;
  };
  analysisDate: string;
}

// Function to get NIB basic information - in a real app this would fetch from API or scrape website
function getNIBBasicInfo(): NIBBasicInfo {
  return {
    yearInBrief: {
      title: "Our Year in Brief",
      description: "NIB is the international financial institution of the Nordic and Baltic countries. Our mission is to finance projects that improve productivity and benefit the environment of the Nordic and Baltic countries.",
      keyHighlights: [
        {
          title: "Lending",
          value: "EUR 5.6 billion",
          description: "NIB financed sustainable growth in its member countries with a total of EUR 5.6 billion in loans, half of them classified with a positive impact on the environment.",
          sourceUrl: "https://www.nib.int/who-we-are/our-year-in-brief"
        },
        {
          title: "Response to Covid-19",
          value: "EUR 1.6 billion",
          description: "Response Loans to alleviate the social and economic impact of the COVID-19 pandemic.",
          sourceUrl: "https://www.nib.int/who-we-are/our-year-in-brief"
        },
        {
          title: "Green Bonds",
          value: "EUR 800 million",
          description: "NIB issued its ninth and largest environmental bond, which was also the Bank's first seven-year EUR benchmark.",
          sourceUrl: "https://www.nib.int/who-we-are/our-year-in-brief"
        }
      ],
      factSheets: [
        {
          title: "Impact Report 2023",
          description: "Our investments and their impact on sustainable development",
          link: "https://www.nib.int/"
        },
        {
          title: "Annual Report 2023",
          description: "Financial performance and key achievements",
          link: "https://www.nib.int/who-we-are/our-year-in-brief"
        },
        {
          title: "NIB Strategy 2023-2025",
          description: "Our strategic framework for delivering on our mission",
          link: "https://www.nib.int/who-we-are/about"
        }
      ]
    },
    quickLinks: [
      {
        title: "What we finance",
        icon: "finance",
        description: "Learn about NIB's financing solutions and eligibility criteria",
        link: "https://www.nib.int/"
      },
      {
        title: "Green Bond Framework",
        icon: "sustainability",
        description: "Our approach to sustainable finance through green bonds",
        link: "https://www.nib.int/"
      },
      {
        title: "Member Countries",
        icon: "globe",
        description: "The Nordic and Baltic countries that own NIB",
        link: "https://www.nib.int/who-we-are/about"
      },
      {
        title: "Projects",
        icon: "projects",
        description: "Browse our financed projects across sectors",
        link: "https://www.nib.int/"
      }
    ]
  };
}

// This function reads NIB data and generates AI-powered investment recommendations
export async function getNIBRecommendations(): Promise<NIBRecommendations> {
  // Get cache service
  const { getCachedData, setCachedData } = await getCacheService();
  
  // Check cache first
  const cacheKey = NIB_RECOMMENDATIONS_CACHE_KEY;
  const cachedData = getCachedData<NIBRecommendations>(cacheKey, 24 * 60 * 60 * 1000); // Cache for 24 hours
  if (cachedData) {
    return cachedData;
  }

  // Fetch basic NIB data
  const basicInfo = getNIBBasicInfo();

  try {
    // Generate AI recommendations for each category
    const sustainableRec = await generateAIRecommendation("sustainable finance", 
      "Generate detailed investment recommendations for sustainable finance and green bonds in the Nordic and Baltic region based on NIB's mandate and focus areas. Include analysis of risks and opportunities.");
    
    const infrastructureRec = await generateAIRecommendation("infrastructure development", 
      "Generate detailed investment recommendations for infrastructure projects in the Nordic and Baltic region based on NIB's mandate and focus areas. Focus on transportation, energy, and digital connectivity. Include analysis of risks and opportunities.");
    
    const innovationRec = await generateAIRecommendation("innovation and technology", 
      "Generate detailed investment recommendations for innovation and technology sectors in the Nordic and Baltic region based on NIB's mandate and focus areas. Focus on digital transformation, clean technology, and bioeconomy. Include analysis of risks and opportunities.");

    // Create complete recommendations object
    const recommendations: NIBRecommendations = {
      basic: basicInfo,
      aiRecommendations: {
        sustainable: sustainableRec,
        infrastructure: infrastructureRec,
        innovation: innovationRec
      },
      analysisDate: new Date().toISOString()
    };

    // Cache the results
    setCachedData(cacheKey, recommendations);

    return recommendations;
  } catch (error) {
    console.error("Error generating NIB AI recommendations:", error);
    // Fallback to basic data if AI generation fails
    return {
      basic: basicInfo,
      aiRecommendations: {
        sustainable: createFallbackRecommendation("Sustainable Finance", "sustainable finance"),
        infrastructure: createFallbackRecommendation("Infrastructure Development", "infrastructure"),
        innovation: createFallbackRecommendation("Innovation & Technology", "innovation")
      },
      analysisDate: new Date().toISOString()
    };
  }
}

// This function uses Mistral AI to generate investment recommendations
async function generateAIRecommendation(sector: string, prompt: string): Promise<AIRecommendation> {
  try {
    // Create system prompt
    const systemPrompt = `You are an expert investment advisor specializing in Nordic and Baltic investments, with deep knowledge of the Nordic Investment Bank (NIB) and its mandate. 
    NIB finances projects that improve productivity and benefit the environment in the Nordic and Baltic countries.

    You will analyze ${sector} investment opportunities in the Nordic and Baltic region.
    
    Provide detailed, practical investment recommendations with:
    1. Title - a concise title for the recommendation
    2. Description - a brief 1-2 sentence description
    3. Industry - the specific industry segment
    4. Risk level - classify as Low, Medium, or High
    5. Opportunity level - classify as Low, Medium, or High
    6. Analysis - a detailed paragraph analyzing the current state and potential
    7. Key recommendation - a concise actionable recommendation
    
    IMPORTANT: You must respond with valid JSON only. Do not include any explanatory text or markdown formatting.
    Your entire response must be a single JSON object with these exact keys:
    {
      "title": "",
      "description": "",
      "industry": "",
      "riskLevel": "",
      "opportunityLevel": "",
      "analysis": "",
      "keyRecommendation": ""
    }
    
    Do not use code blocks, do not use backticks, and do not include any additional text before or after the JSON.
    `;

    // User prompt
    const userPrompt = prompt;

    // Prepare messages for OpenRouter API
    const messages: OpenRouterMessage[] = [
      {
        role: "system",
        content: systemPrompt
      },
      {
        role: "user",
        content: userPrompt
      }
    ];

    // Call OpenRouter API (via Mistral model)
    const response = await fetch("https://openrouter.ai/api/v1/chat/completions", {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Authorization": `Bearer ${process.env.OPENROUTER_API_KEY}`
      },
      body: JSON.stringify({
        model: "mistralai/mistral-small-3.1-24b-instruct:free",
        messages: messages,
        response_format: { type: "json_object" }
      })
    });

    if (!response.ok) {
      const errorText = await response.text();
      console.error(`OpenRouter API error: ${response.status} ${errorText}`);
      throw new Error(`OpenRouter API error: ${response.status}`);
    }

    const openRouterResponse = await response.json() as OpenRouterResponse;
    
    // Parse the JSON from the response
    let content = openRouterResponse.choices[0].message.content;
    console.log("AI Recommendation raw response:", content);
    
    try {
      // Handle the case where the response might be wrapped in markdown code blocks
      // First, try to extract content between ```json and ``` markers
      const jsonBlockMatch = content.match(/```json\s*([\s\S]*?)\s*```/);
      if (jsonBlockMatch && jsonBlockMatch[1]) {
        content = jsonBlockMatch[1];
      } else if (content.includes('```')) {
        // Try to extract between any code block markers
        const codeBlockMatch = content.match(/```\s*([\s\S]*?)\s*```/);
        if (codeBlockMatch && codeBlockMatch[1]) {
          content = codeBlockMatch[1];
        }
      }
      
      // If we get an array instead of a single object, take the first item
      if (content.trim().startsWith('[') && content.trim().endsWith(']')) {
        try {
          const parsedArray = JSON.parse(content);
          if (Array.isArray(parsedArray) && parsedArray.length > 0) {
            return parsedArray[0] as AIRecommendation;
          }
        } catch (e) {
          // If array parsing fails, continue with the original approach
        }
      }
      
      // Sometimes the model adds extra text before or after the JSON
      // Try to extract just the JSON object part
      const jsonMatch = content.match(/\{[\s\S]*?\}/);
      if (jsonMatch) {
        content = jsonMatch[0];
      }
      
      const recommendationData = JSON.parse(content);
      return recommendationData as AIRecommendation;
    } catch (parseError) {
      console.error("Error parsing AI recommendation JSON:", parseError);
      throw new Error("Failed to parse AI response");
    }
  } catch (error) {
    console.error(`Error generating AI recommendation for ${sector}:`, error);
    return createFallbackRecommendation(
      sector.charAt(0).toUpperCase() + sector.slice(1), 
      sector
    );
  }
}

// Create a fallback recommendation if AI generation fails
function createFallbackRecommendation(title: string, sector: string): AIRecommendation {
  return {
    title: title,
    description: `Investment opportunities in ${sector} across Nordic and Baltic countries.`,
    industry: sector,
    riskLevel: "Medium",
    opportunityLevel: "Medium",
    analysis: `Based on NIB's mandate, ${sector} represents a key area of focus with both challenges and opportunities. The Nordic and Baltic region continues to see development in this sector.`,
    keyRecommendation: `Consider targeted investments in ${sector} projects that align with NIB's dual mandate of productivity enhancement and environmental sustainability.`
  };
}